### PR TITLE
RPC Service Architecture Refactor

### PR DESCRIPTION
**Pull request by haikalpribadi**
## What is the goal of this PR?

Over time, our Client architecture, such as Client Java, has been extended considerably to implement various features and optimisations. It has lead the architecture to a complex codebase which is not so easy to follow and replicate in other languages. We have now refactored Client Java considerably to come up with an architecture that is easy to understand, easy to extend, and easy to replicate in other languages that wish to implement a Grakn Client. These changes on the client requires changes in the protocol that we need to apply onto the server-side RPC Service architecture.

### CODE REVIEW

File: migrator/proto/data.proto [hunk 1]
```diff
@@ -59,72 +59,72 @@ option java_outer_classname = "DataProto";
...
+        string id = 1;
+    }
 
-  message Header {
-    string grakn_version = 1;
-    string original_database = 2; // formerly keyspace
-  }
+    message Header {
+        string grakn_version = 1;
+        string original_database = 2; // formerly keyspace
```
Comment (alexjpwalker) [inline migrator/proto/data.proto:108]: We don't need this comment anymore right?
```suggestion
        string original_database = 2;
```

File: server/common/ResponseBuilder.java [hunk 2]
```diff
@@ -143,34 +310,33 @@ public static StatusRuntimeException exception(Throwable e) {
 
     public static class Concept {
 
-        public static ConceptProto.Concept concept(grakn.core.concept.Concept concept) {
+        public static ConceptProto.Concept protoConcept(grakn.core.concept.Concept concept) {
```
Comment (alexjpwalker) [inline server/common/ResponseBuilder.java:313]: Why are these renamed to `protoConcept`, etc? The `proto` prefix seems redundant given this file's whole purpose is to create proto objects.

File: server/query/QueryService.java [hunk 3]
```diff
@@ -98,95 +102,66 @@ public void execute(Transaction.Req request) {
...
+    private void match(String queryStr, Options.Query options, TransactionProto.Transaction.Req eq) {
         GraqlMatch query = Graql.parseQuery(queryStr).asMatch();
         Context.Query context = new Context.Query(transactionSrv.context(), options.query(query), query);
         FunctionalIterator<ConceptMap> answers = queryMgr.match(query, context);
-        transactionSrv.stream(answers, request.getId(), context.options(), as -> response(
-                request, QueryProto.Query.Res.newBuilder().setMatchRes(
-                        QueryProto.Query.Match.Res.newBuilder().addAllAnswers(
-                                iterate(as).map(ResponseBuilder.Answer::conceptMap).toList()))
-        ));
+        transactionSrv.stream(answers, eq.getReqId(), context.options(), a -> matchResPart(eq.getReqId(), a));
```
Comment (alexjpwalker) [inline server/query/QueryService.java:123]: Should we refactor `TransactionService.stream` to not take in a lambda and instead use `map` or `flatMap`?

File: server/common/ResponseBuilder.java (hunk 2 above)
Comment (haikalpribadi) [reply server/common/ResponseBuilder.java:313]: We're doing this because this method is quite exceptional in that the input type, return type, and method name, are all the same word "concept" - it got kind ambiguous.

File: server/query/QueryService.java (hunk 3 above)
Comment (haikalpribadi) [reply server/query/QueryService.java:123]: I tried - it's not possible because we have to consume the iterator based on a time window first, before applying the lambda to create a "ResPart".

**Review by alexjpwalker — APPROVED** [review]
